R/class-plotdata.R

```diff
@@ -89,7 +98,80 @@ newPlotdata <- function(.dt = data.table(),
+    listVariable <- validateListVar(listVariable)
+
+    # Set variable, value names appropriately
+    if(is.null(unique(listVarDetails$inferredVariable$entityId))) {
+      variable.name <- listVarDetails$listVarPlotRef
+      value.name <- listVarDetails$inferredVariable$variableId
+    } else {
+      variable.name <- paste(unique(listVarDetails$inferredVariable$entityId),listVarDetails$listVarPlotRef, sep='.')
+      value.name <- paste(unique(listVarDetails$inferredVariable$entityId),listVarDetails$inferredVariable$variableId, sep='.')
+    }
+
+    # Reshape data
+    .dt <- data.table::melt(.dt, measure.vars = toColNameOrNull(listVariable),
+                        variable.factor = FALSE,
+                        variable.name= variable.name,
+                        value.name=value.name)
+
+    # Assign new variable details for the created categorical variable
+    newCatVariable <- list('variableId' = listVarDetails$listVarPlotRef,
+                   'entityId' = unique(listVariable$entityId),
+                   'dataType' = 'STRING',
+                   'dataShape' = 'CATEGORICAL',
+                   'displayLabel' = listVarDetails$listVarDisplayLabel)
+
+    if (listVarDetails$listVarPlotRef == 'xAxisVariable') {
+      xAxisVariable <- newCatVariable
+      x <- toColNameOrNull(xAxisVariable)
+      xType <- emptyStringToNull(as.character(xAxisVariable$dataType))
+      xShape <- emptyStringToNull(as.character(xAxisVariable$dataShape))
+      .dt[[x]] <- updateType(.dt[[x]], xType, xShape)
+
+    } else if (listVarDetails$listVarPlotRef == 'overlayVariable') {
+      overlayVariable <- newCatVariable
+      group <- toColNameOrNull(overlayVariable)
+      groupType <- emptyStringToNull(as.character(overlayVariable$dataType))
+      groupShape <- emptyStringToNull(as.character(overlayVariable$dataShape))
+      .dt[[group]] <- updateType(.dt[[group]], groupType, groupShape)
+
+    } else if (listVarDetails$listVarPlotRef == 'facetVariable1') {
+      facetVariable1 <- newCatVariable
+      facet1 <- toColNameOrNull(facetVariable1)
+      facetType1 <- emptyStringToNull(as.character(facetVariable1$dataType))
+      facetShape1 <- emptyStringToNull(as.character(facetVariable1$dataShape))
+      .dt[[facet1]] <- updateType(.dt[[facet1]], facetType1, facetShape1) 
+
+      panelData <- makePanels(.dt, facet1, facet2)
+      .dt <- data.table::setDT(panelData[[1]])
+      panel <- panelData[[2]]
+    }
+
+    # Assume inferredVarPlotRef = yAxisVariable always.
+    yAxisVariable <- listVarDetails$inferredVariable
+    y <- toColNameOrNull(yAxisVariable)
+    yType <- emptyStringToNull(as.character(yAxisVariable$dataType))
+    yShape <- emptyStringToNull(as.character(yAxisVariable$dataShape))
+    .dt[[y]] <- updateType(.dt[[y]], yType, yShape) 
+
+    data.table::setcolorder(.dt, c(x, y, z, group, panel))
+
+  }
+  
+  # If overlay is continuous and NOT a listvar, it does not contribute to final groups
```

Comment:
I'd love to do that. The reason the reshaping comes at the end is so that the supporting tables (ex. complete cases) work out nicely. But I'll have a go at seeing if we really need to updateType at the beginning or if that can get moved further down.